### DISCUSSION TITLE
Does any CoreDNS plugin supports forwarding query to be answered by a DoH?

### DISCUSSION THREAD

**Discussion by xinbenlv** [coredns/coredns#6889] [Ideas]
Instead of listen to incoming query as a http server using DoH, I what I am referring to is 
when CoreDNS instance received a DNS query, it make an HTTP request to a pre-configured endpoint 

for example, here is one possible setup

`ns1.example-coredns.com` is the nameserver, receiving a query for "abc.com" by `dig @ns1.example-coredns.com abc.com`

then behind the scene, ns1.example-coredns.com was configured in the Corefile that 

```
abc.com {
  doh: http://customized-ns.com/dns-query
}
```

So it makes a POST query to http://customized-ns.com/dns-query for abc.com as a DoH looklup, and got a response, then respond to the client.

Does such plugin exist, or if we contribute to build this, do you think it's worth to be included in the core plugins?

Thank you!

**Comment by johnbelamaric** [comment]
Forward accepts TLS as the protocol, I think that’s what you want: https://coredns.io/plugins/forward/

**Comment by xinbenlv** [comment]
Thanks @johnbelamaric. I am aware of `forward` plugin, I think the TLS support has several shortcoming that doesn't match what a HTTP / HTTPS can support.
1. TLS in `forward` currently requires IP address and port, and HTTP / HTTPS allows you to use a domain and URL which is much more convenient and widely supported. For example in our use case we might want to use load balance
2. TLS in `forward` also requires setting up certificates and keys, HTTP/HTTPS can abstract out the complications.

My two cents that's why I want to find/ propose forward with HTTP / HTTPS endpoint or a new plugin

**Comment by xinbenlv** [comment]
So far it seems there is no known plugin. (and please correct me if I am wrong, we don't want to reinvent the wheel) 
We are going to proceed with building one for ourselves as a thirdparty plugin, if the CoreDNS community believe this will be beneficial, we want to contribute it into the CoreDNS core plugins.

The development will be tracked in #6890

**Reply by chrisohaver** [reply]
There is an external plugin.

https://github.com/v-byte-cpu/coredns-https

The author has not added it to the list of external plugins on coredns.io.

It is based on a coredns/coredns PR from a few years ago that was ultimately not merged.

**Comment by xinbenlv** [comment]
This is great, thank you @chrisohaver I will take a look and see if it works. Do you know what was the rationale that it didn't get listed as external plugins on coredns.io?

**Comment by chrisohaver** [comment]
I don’t recall a PR to add it to coredns.io.
It’s up to the maintainer of the external plugin to open a PR for adding it to coredns.io.  

As for getting merged into coredns in the first place as an internal plugin, there was a difference of opinions and ultimately the PR was not completed by the original author to the satisfaction of the reviewer.
